client/ping.go

```diff
@@ -71,6 +74,38 @@ func parsePingResponse(cli *Client, resp serverResponse) (types.Ping, error) {
 			ControlAvailable: role == "manager",
 		}
 	}
+	if engineFeatures := resp.header.Get("Engine-Features"); engineFeatures != "" {
+		featuresMap, err := parseEngineFeaturesHeader(engineFeatures)
+		if err != nil {
+			return ping, err
+		}
+		ping.EngineFeatures = featuresMap
+	}
 	err := cli.checkResponseErr(resp)
 	return ping, errdefs.FromStatusCode(err, resp.statusCode)
 }
+
+// Prevent a malicious engine from blowing up clients by sending
+// a huge number of engine features.
+const maxNumEngineFeatures = 100
+
+func parseEngineFeaturesHeader(header string) (map[string]bool, error) {
+	if numFeatures := strings.Count(header, ",") + 1; numFeatures > maxNumEngineFeatures {
+		return nil, errors.Errorf("too many features in Engine-Features header – expected max %d, found %d", maxNumEngineFeatures, numFeatures)
+	}
+
+	features := strings.Split(header, ",")
+	featuresMap := make(map[string]bool)
+	for _, featureValue := range features {
+		parts := strings.Split(featureValue, "=")
```

Comment:
This can be handled by `strings.Cut`, which also has less / no allocations;

```go
k, v, found := strings.Cut(featureValue, "=")
if !found {
    return nil, fmt.Errorf("failed to parse Engine-Features header feature (%s): missing '='", featureValue)
}
```

Reply:
Sounds good, thanks!